File: Ferrum/Errors/BaseError.fs
namespace Ferrum

open System


[<AbstractClass>]
type BaseError() =

    abstract Message: string
    abstract InnerError: IError voption

    interface IError with
        member this.InnerError = this.InnerError
        member this.Message = this.Message

    interface IFormattable with
        member this.ToString(format, _formatProvider) =
            let formatter = ErrorFormatters.getByFormat format
            formatter.Format(this)

    override this.ToString() =
        ErrorFormatters.General.Format(this)

File: Ferrum/Errors/BaseError.cs
using Ferrum.Formatting;

namespace Ferrum.Errors;

public abstract class BaseError : IError, IFormattable
{
    public abstract string Message { get; }
    public abstract IError? InnerError { get; }

    public string ToString(string? format, IFormatProvider? formatProvider)
    {
        return (format is not null ? ErrorFormatter.ByFormat(format) : ErrorFormatter.Default).Format(this);
    }

    public override string ToString()
    {
        return ErrorFormatter.Default.Format(this);
    }
}
